torchvision/csrc/io/decoder/gpu/gpu_decoder.cpp

```diff
@@ -0,0 +1,85 @@
+#include "gpu_decoder.h"
+#include <c10/cuda/CUDAGuard.h>
+
+/* Set cuda device, create cuda context and initialise the demuxer and decoder.
+ */
+GPUDecoder::GPUDecoder(std::string src_file, int64_t dev)
+    : demuxer(src_file.c_str()), device(dev) {
+  at::cuda::CUDAGuard device_guard(device);
+  check_for_cuda_errors(
+      cuDevicePrimaryCtxRetain(&ctx, device), __LINE__, __FILE__);
+  decoder.init(ctx, ffmpeg_to_codec(demuxer.get_video_codec()));
+  initialised = true;
+}
+
+GPUDecoder::~GPUDecoder() {
+  at::cuda::CUDAGuard device_guard(device);
+  decoder.release();
+  if (initialised) {
+    check_for_cuda_errors(
+        cuDevicePrimaryCtxRelease(device), __LINE__, __FILE__);
+  }
+}
+
+/* Fetch a decoded frame tensor after demuxing and decoding.
+ */
+torch::Tensor GPUDecoder::decode() {
+  torch::Tensor frameTensor;
+  unsigned long videoBytes = 0;
+  uint8_t* video = nullptr;
+  at::cuda::CUDAGuard device_guard(device);
+  auto options = torch::TensorOptions().dtype(torch::kU8).device(torch::kCUDA);
+  torch::Tensor frame = torch::zeros({0}, options);
```

Comment:
nit: I believe you can just do
```c
torch::Tensor frame;
```
now that `decoder.fetch_frame()` returns a tensor of the right dtype and device

Reply:
https://github.com/pytorch/vision/issues/5148